svc/job/eta.go

```diff
@@ -0,0 +1,120 @@
+package job
+
+import (
+	"context"
+	"fmt"
+	"time"
+
+	"github.com/openmarketplaceengine/geoservices"
+	"github.com/openmarketplaceengine/geoservices/distance"
+	"github.com/openmarketplaceengine/geoservices/distance/google"
+	"github.com/openmarketplaceengine/openmarketplaceengine/cfg"
+	"github.com/openmarketplaceengine/openmarketplaceengine/dao"
+	"github.com/openmarketplaceengine/openmarketplaceengine/dom/job"
+	"github.com/openmarketplaceengine/openmarketplaceengine/svc/location"
+	"googlemaps.github.io/maps"
+)
+
+type Estimate struct {
+	DistanceMeters int
+	Duration       time.Duration
+}
+
+type EstimatedJob struct {
+	ID              dao.SUID
+	WorkerToPickup  Estimate
+	PickupToDropOff Estimate
+	WorkerLocation  Location
+	PickupLocation  Location
+	DropOffLocation Location
+}
+
+type Location struct {
+	Address string
+	Lat     float64
+	Lng     float64
+}
+
+func estimatedJobs(ctx context.Context, workerLocation *location.LastLocation, jobs []*job.AvailableJob) ([]*EstimatedJob, error) {
+	if jobs == nil {
+		return nil, fmt.Errorf("jobs argument cannot be nil")
+	}
+	if len(jobs) == 0 {
+		return nil, fmt.Errorf("jobs argument cannot be empty")
+	}
+
+	client, err := maps.NewClient(maps.WithAPIKey(cfg.Server.GoogleAPIKey))
+	if err != nil {
+		return nil, fmt.Errorf("maps client create error: %w", err)
+	}
+
+	origins, destinations := transform(jobs)
+	originsPlus := append(origins, geoservices.LatLng{Lat: workerLocation.Latitude, Lng: workerLocation.Longitude})
+	m, err := google.Matrix(ctx, client, distance.MatrixPointsInput{
+		Origins:      originsPlus,
+		Destinations: destinations,
+	})
+
+	if err != nil {
+		return nil, fmt.Errorf("google matrix API error: %w", err)
+	}
+
+	totalOrigins := len(jobs) + 1
+	if len(m.Rows) != totalOrigins {
+		return nil, fmt.Errorf("google matrix API expected to return exactly %v rows because of +1 worker origin, but got %v", totalOrigins, len(m.Rows))
+	}
```

Comment:
nit: Not 100% convinced we need this.

Reply:
Also, there may be a bug here? This is the function that maps 1 Worker to N Jobs. The origin here is the worker. So shouldn't the number of origins always be 1?

Reply:
here we receive in one call ETAs for pairs driver-2-pickup pickup-2-dropoff.

for each origin in list, google will return matrix with ETAs to each destination. If fact, the way we do this, we get even more than we need, such as each pickup to each dropoff pairs.(and we need to take proper pairs, ignoring the rest )

we form origins[] and destinations[] arrays. then append to origins[] driver location as last item. 
google return array for each origin element we have array of destinations. The last index always will be driver-2-pickup ETA.

The only task for us is to handle response properly paying attention to these pairs corresponding indexes when we iterate over multi dimensional array(s).

`origins, destinations := transform(jobs)
	originsPlus := append(origins, geoservices.LatLng{Lat: workerLocation.Latitude, Lng: workerLocation.Longitude})
	m, err := google.Matrix(ctx, client, distance.MatrixPointsInput{
		Origins:      originsPlus,
		Destinations: destinations,
	})

	if err != nil {
		return nil, fmt.Errorf("google matrix API error: %w", err)
	}

	totalOrigins := len(jobs) + 1
	if len(m.Rows) != totalOrigins {
		return nil, fmt.Errorf("google matrix API expected to return exactly %v rows because of +1 worker origin, but got %v", totalOrigins, len(m.Rows))
	}

	workerIdx := len(m.Rows) - 1
	workerRow := m.Rows[workerIdx]

	var eJobs = make([]*Job, len(jobs))
	for i, row := range m.Rows {
		if i == workerIdx {
			break
		}
		for j, e := range row.Elements {
			eJobs[j] = &Job{
				ID: jobs[j].ID,
				WorkerLocation: Location{
					Address: m.OriginAddresses[workerIdx],
					Lat:     workerLocation.Latitude,
					Lng:     workerLocation.Longitude,
				},
				PickupLocation: Location{
					Address: m.OriginAddresses[j],
					Lat:     origins[j].Lat,
					Lng:     origins[j].Lng,
				},
				DropOffLocation: Location{
					Address: m.DestinationAddresses[j],
					Lat:     destinations[j].Lat,
					Lng:     destinations[j].Lng,
				},
				WorkerToPickup: Estimate{
					DistanceMeters: workerRow.Elements[j].Distance,
					Duration:       workerRow.Elements[j].Duration,
				},
				PickupToDropOff: Estimate{
					DistanceMeters: e.Distance,
					Duration:       e.Duration,
				},
			}
		}
	}
`

Reply:
I think the 1st request to GMaps should be:
**1 driver to 25 job locations**

I think `PickupToDropOff` should stay nil for now until we settle on a strategy, but naively we could follow-up with up to 3 requests to Google:
1. `[J1.Pickup, ..., J10.Pickup] x [J1.Dropoff, ..., J10.Dropoff]`
2. `[J11.Pickup, ..., J20.Pickup] x [J11.Dropoff, ..., J20.Dropoff]`
3. `[J21.Pickup, ..., J25.Pickup] x [J21.Dropoff, ..., J25.Dropoff]`

Reply:
100 matrix elements total, is something i totally missed :-(
yep, we are exceeding this in current implementation in the case of 25 jobs at input. 

Need to optimise/divide requests, etc. 

@kevinmichaelchen, do you prefer fixing it in current PR? (i prefer separate PR, as this one is getting quite big already)

Reply:
I think if we ignore `PickupToDropOff` for now we should be okay. You can handle those in a later PR.